ComponentKit/Core/Scope/CKDetectComponentScopeCollisions.mm

```diff
@@ -31,9 +31,53 @@
 }
 #pragma clang diagnostic pop
 
-void CKDetectComponentScopeCollisions(const CKComponentLayout &layout)
+static CKComponent *completeComponentScopeCollisionPair(CKComponent *collidingComponent,
+                                            id<NSObject> collidingScope,
+                                            NSMapTable<CKComponent *, CKComponent*> *componentsToParentComponents)
+{
+  for (CKComponent* componentKey in componentsToParentComponents) {
+    const id<NSObject> scopeFrameToken = [componentKey scopeFrameToken];
+    if ([collidingScope isEqual:scopeFrameToken] && ![collidingComponent isEqual:componentKey]) {
+      return componentKey;
+    }
+  }
+  return nil;
+}
+
+static CKComponent *lowestCommonAncestor(CKComponent *component,
+                                         id<NSObject> collisionScope,
+                                         NSMapTable<CKComponent *, CKComponent*> *componentsToParentComponents)
+{
+  // First we need to find the node we're having a collision with.
+  CKComponent *collidingComponent = completeComponentScopeCollisionPair(component, collisionScope, componentsToParentComponents);
+  if ([collidingComponent isEqual:component]) {
+    return [componentsToParentComponents objectForKey:component];
```

Comment:
So, first scenario: the colliding component is the same as the component. Return the parent in that case.

When would this situation arise? It feels a bit unlikely but definitely worth covering this situation. Maybe add some comments explaining this base case?